connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala

```diff
@@ -38,79 +38,77 @@ private[kusto] case class KustoRelation(kustoCoordinates: KustoCoordinates,
+      kustoCoordinates.ingestionUrl, kustoCoordinates.clusterAlias)
+    val isUserOptionForceSingleMode = readOptions.readMode.contains(ReadMode.ForceSingleMode)
+    val storageParameters = maybeStorageParameters.getOrElse(kustoClient.getTempBlobsForExport)
+    val (useSingleMode, estimatedRecordCount) = readOptions.readMode match {
+      // if the user provides a specific option , this is to be used no matter what
+      case Some(_) => (isUserOptionForceSingleMode, -1)
+      // If there is no option mentioned , then estimate which option to use
+      // Count records and see if we wat a distributed or single mode
+      case None => val estimatedRecordCountResult =  Try(KDSU.
+        estimateRowsCount(kustoClient.engineClient, query, kustoCoordinates.database, clientRequestProperties.orNull))
+        estimatedRecordCountResult match {
+          // if the count is lss than the 5k threshold,use single mode.
+          case Success(recordCount) => (recordCount <= KustoConstants.DirectQueryUpperBoundRows, recordCount)
+          // A case of query timing out. ForceDistributedMode will be used here
+          case Failure(_) => (false, -1)
+        }
     }
-
-    var exception: Option[Exception] = None
-    if(res.isEmpty) {
-      if (useSingleMode) {
-        try {
-          res = Some(KustoReader.singleBuildScan(
-            kustoClient,
-            KustoReadRequest(sparkSession, cachedSchema, kustoCoordinates, query, authentication, timeout, clientRequestProperties, requestId),
-            KustoFiltering(requiredColumns, filters)))
-        } catch {
-          case ex: Exception => exception = Some(ex)
+    // if the selection yields 0 rows there are no records to scan and process
+    if(estimatedRecordCount == 0){
+      sparkSession.emptyDataFrame.rdd
+    } else {
+      // either a case of non-zero records or a case of timed-out.`
+      if(useSingleMode){
+        // there are less than 5000 (KustoConstants.DirectQueryUpperBoundRows) records, perform a single scan
+        Try(KustoReader.singleBuildScan(
+          kustoClient,
+          KustoReadRequest(sparkSession, cachedSchema, kustoCoordinates, query, authentication, timeout, clientRequestProperties, requestId),
+          KustoFiltering(requiredColumns, filters))) match {
+          case Success(rdd) => rdd
+          case Failure(exception) =>
+            KDSU.logError("KustoRelation","Failed with Single mode, falling back to Distributed mode," +
+              s"requestId: $requestId. Exception : ${exception.getMessage}")
+            // If the user specified forceSingleMode explicitly and that cannot be honored , throw an exception back
+            if(isUserOptionForceSingleMode){
+              // Expected behavior for Issue#261
+              throw exception
+            } else {
+              // The case where used did not provide an option and we estimated to be a single scan.
```

Comment:
add 
TODO: Parse exception. Fallback only if failed on 'too many rows'.

Reply:
Added this parsing logic , In case that the exception on ForceSingleScan is too many rows, we fallback to distributed mode. All other exceptions for ForceSingleMode get thrown back and do not fallback